core/src/main/java/com/linecorp/armeria/common/util/DefaultAsyncLoader.java

```diff
@@ -0,0 +1,255 @@
+                }
+            }
+
+            future = new RefreshingFuture<>(cacheEntry);
+            if (loadFutureUpdater.compareAndSet(this, loadFuture, future)) {
+                break;
+            }
+        }
+
+        final T cache = cacheEntry != null ? cacheEntry.value : null;
+        load(cache, future);
+
+        return future;
+    }
+
+    private void load(@Nullable T cache, CompletableFuture<CacheEntry<T>> future) {
+        try {
+            requireNonNull(loader.apply(cache), "loader.apply() returned null")
+                    .handle((value, cause) -> {
+                        if (cause != null) {
+                            logger.warn("Failed to load a new value from loader: {}. cache: {}",
+                                        loader, cache, cause);
+                            handleException(cause, cache, future);
+                        } else {
+                            future.complete(new CacheEntry<>(value));
+                        }
+                        return null;
+                    });
+        } catch (Exception e) {
+            logger.warn("Unexpected exception from loader.apply()", e);
+            handleException(e, cache, future);
+        }
+    }
+
+    private boolean needsRefresh(CacheEntry<T> cacheEntry) {
+        boolean refresh = false;
+        final T cache = cacheEntry.value;
+        try {
+            refresh = refreshIf != null && refreshIf.test(cache);
+        } catch (Exception e) {
+            logger.warn("Unexpected exception from refreshIf.test()", e);
+        }
+
+        return refresh;
+    }
+
+    private void handleException(Throwable cause, @Nullable T cache,
+                                 CompletableFuture<CacheEntry<T>> future) {
+        if (exceptionHandler == null) {
+            future.completeExceptionally(cause);
+            return;
+        }
+
+        try {
+            final CompletableFuture<T> fallback = exceptionHandler.apply(cause, cache);
+            if (fallback == null) {
+                future.completeExceptionally(cause);
+                return;
+            }
+
+            fallback.handle((value, cause0) -> {
+                if (cause0 != null) {
+                    logger.warn("Failed to load a new value from exceptionHandler: {}. " +
+                                "cache: {}", exceptionHandler, cache, cause0);
+                    future.completeExceptionally(cause0);
+                } else {
+                    future.complete(new CacheEntry<>(value));
+                }
+                return null;
+            });
+        } catch (Exception e) {
+            logger.warn("Unexpected exception from exceptionHandler.apply()", e);
+            future.completeExceptionally(cause);
+        }
+    }
+
+    private boolean isValid(@Nullable CacheEntry<T> cacheEntry) {
+        if (cacheEntry == null) {
+            return false;
+        }
```

Comment:
Why is `null` an invalid entry value? A user might want to cache `null` as the result.

Reply:
> A user might want to cache null as the result.

I guess `cacheEntry.value` would be `null` in this case

Reply:
```java
    private volatile RefreshingFuture<T> loadFuture = RefreshingFuture.completedFuture(); // complete(null);
```
- `CacheEntry` is `null` only when there's nothing cached. (e.g. right after AsyncLoader is initialized)
- When user want to cache null result, `CacheEntry.value` is `null`

Reply:
My question is: Let's say the loader function returns `null` as the result of asynchronous computation. Is there a way for this PR to distinguish between 1) value not loaded and 2) value loaded and is null?

`null` isn't necessarily an invalid value. Therefore, the following sentence is not always true:

> CacheEntry is null only when there's nothing cached.

Reply:
```java
    private void load(@Nullable T cache, CompletableFuture<CacheEntry<T>> future) {
        try {
            requireNonNull(loader.apply(cache), "loader.apply() returned null")
```
Aha~ I understand what you saying. we previously decided that "loader should not return null" , so checking like above.

```java
     * @param loader function to load value. {@code T} is previously cached value
     *               or {@code null} when nothing is cached.
```
If null is returned, currently is means nothing is cached yet.

> Is there a way for this PR to distinguish between 1) value not loaded and 2) value loaded and is null?

Currently we don't have a way to distinguish it.
If we allow user to "return null on loader.apply() and cache it", we need to fix several logics. WDYT?

- If nothing is cached yet, we need a way to distinguish it with `null` is cached
- On `isValid()`, we need to handle `null` is cached by user case
- ..

Reply:
> A user might want to cache null as the result.

I think it's also good to allow user to cache null. If you want, please tell me~! I'll update PR 😄

Reply:
Yes, please. Let's allow a user to cache `null`.

Reply:
Fixed.